### DISCUSSION TITLE
ubuntuへ移行してからのcondaについて

### DISCUSSION THREAD

**Discussion by Tomohiro-gh** [nig-sc/nigsc_homepage2#419] [Q&A] [answered]
お世話になっております．

シェルスクリプトへ conda環境を記載したジョブを投入したところ，condaでエラーが出てしまいました．
OS移行前はこれで動いていたのですが，何か変更点などあるでしょうか？

スクリプト (veloという環境を作成した)
```
conda activate velo
```

エラー内容 >>>
CondaError: Run 'conda init' before 'conda deactivate'

Could not find conda environment: velo
You can list all discoverable environments with `conda info --envs`.
<<< 

申し訳ありませんがご教示いただければ幸いです．
よろしくお願い致します．

**Comment by sc-fs** [comment]
一つ、試していただきたいのですが。
実行スクリプトに「#$ -V」があるようでしたら、一旦コメントアウトして
対象JOBを再実行していただけますでしょうか。
よろしくお願いします。

**Comment by Tomohiro-gh** [comment]
ありがとうございます．スクリプトに -Vオプションありましたのでのぞいてみたのですが，同様のエラーが出てしまいます．
コマンドに conda activate veloと直接打てば環境に入れるのですが．．．
お手数ですが他に試せることあれば教えていただければ幸いです．
よろしくお願い致します．

----- 下記エラーになります．

Could not find conda environment: velo
You can list all discoverable environments with `conda info --envs`.

/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found
/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found
/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found
/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found
/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found
/var/spool/age/m09/job_scripts/24986443: line 64: velocyto: command not found

**Reply by sc-fs** [reply]
動作中に意図しない環境になり、qsubされたかと思い、環境を引きずらないよう
「#$ -V」をコメントアウトして実行して頂きましたが、同じ状況のようですので、
.bashrcに設定されているであろう、「conda initialize」が動作していないのかもしれません。

お手数ですが、もう一つ、試していただきたいのですが、
実行JOBの先頭で「source  ~/.bashrc」を記載して再度、qsub実施していただけないでしょうか。

=====（JOB抜粋）
```
#$ -e velocyto_MuscleIntestine_240113.Err.txt

source  ~/.bashrc
# working directory
```
=====

よろしくお願いします。

**Reply by Tomohiro-gh** [reply]
お返事遅れてすいません．
ご教示いただきありがとうございます．

先ほどやってみましたが，やはりダメでした．
下記のようにエラーが出てしまいます．

Could not find conda environment: velo
You can list all discoverable environments with `conda info --envs`.

/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found
/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found
/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found
/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found
/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found
/var/spool/age/m08/job_scripts/24993618: line 62: velocyto: command not found

お手数おかけして申し訳ありません．

**Comment by oogasawa** [comment]
CentOS用のconda環境と、Ubuntu Linux用のConda環境を作って、切り替えるという使い方になると思います。

CentOSで作ったこれまでのconda環境とは別に
Ubuntu Linux用のconda環境を以下のように作ります、`conda create -p $HOME/ubuntu_env python=3.x`  

それで、環境に応じてconda activateすればよいと思います。
異なるOSでconda環境を混ぜて使うとOS固有のライブラリ依存関係の違いにより問題が発生する可能性が高いです。

**Reply by Tomohiro-gh** [reply]
ありがとうございます．
実はOSを移行してから anaconda3を入れ直し，新しくconda環境を使ってやっておりました．
他に解決策があればご教示いただければ幸いです．
よろしくお願い致します．

**Comment by sc-fs** [comment, marked as answer]
今回発生したcondaの初期化が出来ていないため、仮想環境への切替が出来ない原因ですが、
ホームにあるシェル設定ファイルの内容によるものようです。
以下にシェル設定ファイルの読み込み順があります。
https://sc.ddbj.nig.ac.jp/faq/faq_software#shell-config

Tomohiroさんのホーム環境に.bash_profile、 .bashrc、.profileがあり、
.bash_profileの最後にcondaの初期化処理の記載があるかと思います。

この環境で、qsubから実行されたジョブは、ホームに.bash_profileがあるので、.bash_profile読み込みます。
記事にもありますが、.bashrc、.profileは読み込みされません。

.bash_profileには　.bashrcを読み込む記載
「`source ~/.bashrc`」になっているかと思いますが、
.bashrcの先頭にある以下で、スクリプトからの起動は以降の設定をせずに.bashrcの読み込みが終わります。
```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```
結果、condaの初期化処理が実施されませんので、仮想環境への切替が出来ない状況になります。

単純な対応としては、
・.bash_profileにcondaの初期化処理を記載する
・.bashrcのIf not running interactively, don't do anythingの前にcondaの初期化処理を記載する
になると思います。

**Comment by sc-fs** [comment]
実際の対応については、
対処方法に記載した２項目の内、どちらかで対応願います。

**Reply by Tomohiro-gh** [reply]
詳細に調べていただきありがとうございました．
よく理解できました．
今回.bash_profileにてcondaが作動したため，このまま使用してみたいと思います．
ありがとうございました．
